src/source/to-shift.js

```diff
@@ -1,31 +1,98 @@
-    delete obj.length;
+    if (!m) {
+        return;
+    }
 
     let { params, body } = obj.contents;
-    delete obj.contents;
 
+    obj.type = m[2] || m[1];
     obj.params = params;
 
-    if (obj.type === 'ArrowExpressionWithExpression') {
-        obj.type = 'ArrowExpression';
-        obj.body = body;
-    } else {
-        if (obj.type === 'ArrowExpressionWithFunctionBody') {
-            obj.type = 'ArrowExpression';
-        }
-        obj.body = {
+    if (m[1] !== 'ArrowExpressionWithExpression') {
+        body = {
             type: 'FunctionBody',
             directives: obj.directives,
             statements: body
         };
     }
+
+    obj.body = body;
 }
 
-function convertObject(key, value) {}
+function convertObject(key, obj) {
+    switch (obj.type) {
+        case 'Block':
+            switch (this.type) {
+                case 'TryCatchStatement':
+                case 'TryFinallyStatement':
+                case 'CatchClause':
+                    // not used as a statement here, do nothing
+                    break;
+                default:
+                    // otherwise wrap
+                    obj = {
+                        type: 'BlockStatement',
+                        block: obj
+                    };
+                    break;
+            }
+            break;
+        case 'VariableDeclaration':
+            if (
+                this.type !== 'Export' &&
+                !(this.type === 'ForStatement' && key === 'init')
+            ) {
+                // used as a statement, wrap
+                obj = {
+                    type: 'VariableDeclarationStatement',
+                    declaration: obj
+                };
+            }
+            break;
+        case 'LabelledStatement':
+            obj.type = 'LabeledStatement';
+            break;
+        case 'LiteralPropertyName':
+            obj.type = 'StaticPropertyName';
+            break;
+        case 'LiteralRegExpExpression': {
+            let flags = obj.flags;
+            obj.global = flags.includes('g');
+            obj.ignoreCase = flags.includes('i');
+            obj.multiLine = flags.includes('m');
+            obj.sticky = flags.includes('y');
+            obj.unicode = flags.includes('u');
+            break;
+        }
+        case 'ForInOfBinding':
+            obj.type = 'VariableDeclaration';
+            obj.declarators = [
+                {
+                    type: 'VariableDeclarator',
+                    init: null,
+                    binding: obj.binding
+                }
+            ];
+            break;
+        default:
+            tryRemoveFunctionContents(obj);
+            break;
+    }
+
+    return obj;
+}
+
+// This is a reviver callback for JSON.parse to convert Binary AST JSON to Shift AST.
+module.exports = function convert(_key, value) {
+    if (typeof value === 'object' && value !== null && !Array.isArray(value)) {
+        return convertObject.apply(this, arguments);
```

Comment:
`convertObject.call(this, _key, value);` maybe? (to clarify this doesn't expect anything else from `arguments`)

Reply:
This one tends to be slightly faster, but it's more of a habit than important difference in this particular case. Will change.